File: EXAM/TREE/BINARY TREE/PRE-ORDER/pre-order.java
// LeetCode
// Binary Tree Preorder Traversal
// https://leetcode.com/explore/learn/card/data-structure-tree/134/traverse-a-tree/928

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> l = new ArrayList<>();
        
        printPreorder(root, l);
        
        return l;
    }
    
    // Node in Pre Order Of Given Binary Tree
    void printPreorder(TreeNode treeNode, List<Integer> l){
            
        if (treeNode != null){

            /* first Add data of node to List<Integer> */
            l.add(treeNode.val);

            /* then recur on left subtree */
            printPreorder(treeNode.left, l);

            /* now recur on right subtree */
            printPreorder(treeNode.right, l);

        }
    }
}

File: EXAM/TREE/BINARY TREE/PRE-ORDER/pre-order.cpp
#include <iostream>
using namespace std;

/* A binary tree node has data, pointer to left child
and a pointer to right child */
struct Node {
    int data;
    struct Node *left, *right;
    Node(int data)
    {
        this->data = data;
        left = right = NULL;
    }
};

void printPreorder(struct Node*ROOT)
{
    if (ROOT != NULL){
        
        /* first print data of node */
        cout << ROOT->data << " ";
        
        /* then recur on left sutree */
        printPreorder(ROOT->left);
 
        /* now recur on right subtree */
        printPreorder(ROOT->right);
    }
};

int main() {
    struct Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    
    printPreorder(root);
    return 0;
}